### PR TITLE
Migrate `nll_loss2d` from TH to ATen (CPU)

### PR DESCRIPTION
**Pull request by andreaskoepf**
Added check for indicies in Reduction::None case.

### Benchmark results

Note: Due to the size of the input tensors this time the random number generation is responsible for a significant portion of the total time. It is better to look at the individual net time-outputs (which do not include the input preparation).
Script used for benchmark.: [nnl_loss2d_benchmark.py](https://gist.github.com/andreaskoepf/5864aa91e243317cb282c1e7fe576e1b) 

#### WITH PR applied
```
using reduction:  none
CPU forward 1000 took 7.916500908322632e-05
CPU forward 10000 took 0.0002642290201038122
CPU forward 100000 took 0.003828087996225804
CPU forward 1000000 took 0.037140720000024885
CPU forward 10000000 took 0.33387596398824826
CPU forward TOTAL time 7.218988707987592

using reduction:  mean
CPU forward 1000 took 9.165197843685746e-05
CPU forward 10000 took 0.0005258890159893781
CPU forward 100000 took 0.0050761590246111155
CPU forward 1000000 took 0.047345594997750595
CPU forward 10000000 took 0.4790863030066248
CPU forward TOTAL time 7.9106070210109465
CPU for- & backward 1000 took 0.0005489500181283802
CPU for- & backward 10000 took 0.0015284279943443835
CPU for- & backward 100000 took 0.015138130984269083
CPU for- & backward 1000000 took 0.15741890601930209
CPU for- & backward 10000000 took 1.6703072849777527
CPU for- & backward TOTAL time 9.555764263990568

using reduction:  sum
CPU forward 1000 took 8.789298590272665e-05
CPU forward 10000 took 0.000514078012201935
CPU forward 100000 took 0.005135576997417957
CPU forward 1000000 took 0.04715992201818153
CPU forward 10000000 took 0.4821214270195924
CPU forward TOTAL time 7.9119505700073205
CPU for- & backward 1000 took 0.00047759301378391683
CPU for- & backward 10000 took 0.0015945070190355182
CPU for- & backward 100000 took 0.018208994006272405
CPU for- & backward 1000000 took 0.15904426100314595
CPU for- & backward 10000000 took 1.5679037219961174
CPU for- & backward TOTAL time 9.495157692988869
```

#### WITHOUT original TH impl
```
using reduction:  none
CPU forward 1000 took 0.0003981560003012419
CPU forward 10000 took 0.0035912430030293763
CPU forward 100000 took 0.035353766987100244
CPU forward 1000000 took 0.3428319719969295
CPU forward 10000000 took 3.364342701010173
CPU forward TOTAL time 11.166179805004504

using reduction:  mean
CPU forward 1000 took 8.63690220285207e-05
CPU forward 10000 took 0.0004704220045823604
CPU forward 100000 took 0.0045734510058537126
CPU forward 1000000 took 0.046232511987909675
CPU forward 10000000 took 0.4191019559802953
CPU forward TOTAL time 7.846049971994944
CPU for- & backward 1000 took 0.0005974550149403512
CPU for- & backward 10000 took 0.0014057719963602722
CPU for- & backward 100000 took 0.013776941981632262
CPU for- & backward 1000000 took 0.13876214998890646
CPU for- & backward 10000000 took 1.3666698939923663
CPU for- & backward TOTAL time 9.10526105100871

using reduction:  sum
CPU forward 1000 took 7.598899537697434e-05
CPU forward 10000 took 0.00046885499614290893
CPU forward 100000 took 0.0044489419960882515
CPU forward 1000000 took 0.04495517900795676
CPU forward 10000000 took 0.418376043002354
CPU forward TOTAL time 7.789334400993539
CPU for- & backward 1000 took 0.0004464260127861053
CPU for- & backward 10000 took 0.0017732900159899145
CPU for- & backward 100000 took 0.01626713399309665
CPU for- & backward 1000000 took 0.11790941300569102
CPU for- & backward 10000000 took 1.4346664609911386
CPU for- & backward TOTAL time 9.294745502003934
```

### CODE REVIEW

**Comment by andreaskoepf** [comment]
references #24741, #24742, #15870,

File: aten/src/ATen/native/LossNLL2d.cpp [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+  check_inputs_nll_loss2d(input, target, weight);
+  const int64_t n_classes = input.size(1);
+  total_weight.resize_({1});
+
+  AT_DISPATCH_FLOATING_TYPES_AND(
+      ScalarType::BFloat16,
+      input.scalar_type(),
+      "nll_loss2d_forward_out_cpu_template",
+      [&] {
+        auto weight_contiguous = weight.contiguous();
```
Comment (ifedan) [inline aten/src/ATen/native/LossNLL2d.cpp]: why do we need to make it contiguous?

Comment (andreaskoepf) [reply aten/src/ATen/native/LossNLL2d.cpp]: I made it contiguous to access it directly via data-pointer. The problem is when using `TensorAccessor` I get an error when I call `access()` when `!weight.defined()` (I think it tries to get the size of dimensions of the tensor which is not possible when it is not defined). I tried that first but I found no really elegant solution that compared to the `weight_data == nullptr ? ...` ternary statements. 
Maybe I could use a pointer to an TensorAccessor but I thought that would be worse than just making weight contiguous and accessing the elements via `weighdata_ptr<scalar_t>()`.

Comment (andreaskoepf) [reply aten/src/ATen/native/LossNLL2d.cpp]: (calling `weight.contiguous();` is possible on undefined tensors)

**Comment by andreaskoepf** [comment]
btw also I created a version which uses parallel_reduce: see [here](https://github.com/andreaskoepf/pytorch/blob/53c0f95d398ab1a65d24adb7bd75f27382fc40d0/aten/src/ATen/native/LossNLL2d.cpp#L152-L185). It is slightly faster, but I am not sure how aggressive the implementations should make use of CPU resources...?

**Comment by andreaskoepf** [comment]
I refactored this PR a bit to match the nll_loss variant: `total_weight` is now accessed via `data_ptr()` instead of TensorAccessor to support scalar inputs for the total_weight tensor and the dispatch bodies have been moved into separate functions to allow building for Windows.

**Comment by andreaskoepf** [comment]
rebased this one

File: aten/src/ATen/native/LossNLL2d.cpp [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    const int64_t batch_size = input.size(0);
+    const int64_t H = input.size(2);
+    const int64_t W = input.size(3);
+
+    output.resize_({batch_size, H, W});
+    auto input_acc = input.accessor<scalar_t, 4>();
+    auto output_acc = output.accessor<scalar_t, 3>();
+    auto target_acc = target.accessor<int64_t, 3>();
+
+    // we check target indicies but cannot throw inside parallel_for
```
Comment (ezyang) [inline aten/src/ATen/native/LossNLL2d.cpp]: nit: indices. Also, I thought `parallel_for` does propagate exceptions correctly; at least, the OpenMP implementation implies that it does error handling:

```
#ifdef _OPENMP
  std::atomic_flag err_flag = ATOMIC_FLAG_INIT;
  std::exception_ptr eptr;
  // choose number of tasks based on grain size and number of threads
  int64_t num_threads = omp_in_parallel() ? 1 : omp_get_max_threads();
  if (grain_size > 0) {
    num_threads = std::min(num_threads, divup((end - begin), grain_size));
  }

#pragma omp parallel num_threads(num_threads)
  {
    int64_t num_threads = omp_get_num_threads();
    int64_t tid = omp_get_thread_num();
    int64_t chunk_size = divup((end - begin), num_threads);
    int64_t begin_tid = begin + tid * chunk_size;
    if (begin_tid < end) {
      try {
        f(begin_tid, std::min(end, chunk_size + begin_tid));
      } catch (...) {
        if (!err_flag.test_and_set()) {
          eptr = std::current_exception();
        }
      } 
    } 
  }     
  if (eptr) {
    std::rethrow_exception(eptr);
  }
#else
  f(begin, end);
#endif
```

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/LossNLL2d.cpp (hunk 2 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/LossNLL2d.cpp]: @ezyang Worker-thread exception propagation is definitely a fine thing to have - sry, I did not check the parallel_for internals! The code in the PR was a modification of the old TH `nll_loss` implementation which I then also used in the 2d version. The range check could and maybe should be implemented with regular `TORCH_CHECK()` inside the `parallel_for` lambda-body.
General question: When the facebook-github-bot 'landing' notification has been posted .. does that mean I must no longer push modifications to this PR but rather wait for the merge to happen and then later create a new PR to address further changes?

Comment (ezyang) [reply aten/src/ATen/native/LossNLL2d.cpp]: It's OK to push new changes, that will just terminate the land and I'll try again later. fbcode master has been sickly so I've been rebasing.

**Pull request closed without merging**